includes/admin/extensions/class-extensions-api.php

```diff
@@ -0,0 +1,207 @@
+				unset( $option['timeout'] );
+
+				return $option;
+			}
+		}
+
+		// Get all of the product data.
+		$all_product_data = $this->get_all_product_data();
+
+		// If no product data was retrieved, let the option sit for an hour.
+		if ( ! $all_product_data ) {
+			$data = array(
+				'timeout' => strtotime( '+1 hour', time() ),
+			);
+			if ( $option && $is_stale ) {
+				$data = array_merge( $option, $data );
+			}
+			update_option(
+				$option_name,
+				$data,
+				false
+			);
+
+			if ( $item_id && ! empty( $option[ $item_id ] ) ) {
+				return is_array( $option[ $item_id ] ) ? $this->product_data->fromArray( $option[ $item_id ] ) : $option[ $item_id ];
+			}
+
+			return $option;
+		}
+
+		$value = array(
+			'timeout' => strtotime( '+1 week', time() ),
+		);
+		if ( $item_id && ! empty( $all_product_data->$item_id ) ) {
+			$item              = $all_product_data->$item_id;
+			$value[ $item_id ] = $this->get_item_data( $item );
+		} elseif ( in_array( $key, array( 'category', 'tag' ), true ) ) {
+			$term_id = $body[ $key ];
+			foreach ( $all_product_data as $item_id => $item ) {
+				if ( 'category' === $key && ( empty( $item->categories ) || ! in_array( $term_id, $item->categories, true ) ) ) {
+					continue;
+				} elseif ( 'tag' === $key && ( empty( $item->tags ) || ! in_array( $term_id, $item->tags, true ) ) ) {
+					continue;
+				}
+				$value[ $item_id ] = $this->get_item_data( $item );
+			}
+		}
+
+		update_option( $option_name, $value, false );
+
+		return $item_id && ! empty( $value[ $item_id ] ) ? $value[ $item_id ] : $value;
+	}
+
+	/**
+	 * Gets all of the product data, either from a tranient or an API request.
+	 *
+	 * @since 2.11.x
+	 * @return object|false
+	 */
+	private function get_all_product_data() {
+		// Possibly all product data is in a transient. If it is, return it.
+		$all_product_data = get_transient( 'edd_all_extension_data' );
+		if ( $all_product_data ) {
+			return $all_product_data;
+		}
+
+		// Otherwise, query the API.
+		$url     = add_query_arg(
+			array(
+				'edd_action' => 'extension_data',
+			),
+			$this->get_products_url()
+		);
+		$request = wp_remote_get(
+			$url,
+			array(
+				'timeout'   => 15,
+				'sslverify' => true,
+			)
+		);
+
+		// If there was an API error, set timeout for 1 hour and return false.
+		if ( is_wp_error( $request ) || ( 200 !== wp_remote_retrieve_response_code( $request ) ) ) {
+			return false;
+		}
+
+		// Fresh data has been retrieved, so populate with fresh data.
+		$all_product_data = json_decode( wp_remote_retrieve_body( $request ) );
+
+		// Store all of the product data for one hour.
+		set_transient( 'edd_all_extension_data', $all_product_data, HOUR_IN_SECONDS );
+
+		return $all_product_data;
+	}
+
+	/**
+	 * Gets the product data as needed for the extension manager.
+	 *
+	 * @since 2.11.x
+	 * @param object $item
+	 * @return ProductData
+	 */
+	private function get_item_data( $item ) {
+		return $this->product_data->fromArray( array(
```

Comment:
@ashleyfae I did this assuming that it would be better, now that we have it, to store the product data object directly instead of storing it as an array. I did want to note that the `$features` property is then saved for every product, even though it isn't populated at this time (and maybe that was wrong), so I did wonder if it would be better to save the data as an array and convert it to the object on retrieval....so theoretically properties that exist in the class but aren't populated for whatever reason won't be saved to the database.

Reply:
Actually I decided to go ahead and do that in 7623366 but can revert it if you think I should.

Reply:
I do prefer saving it as an array and converting on retrieval. :+1: